qme/exceptions.py

```diff
@@ -0,0 +1,177 @@
+
+class NoTasksError(RuntimeError):
+    """Thrown if tasks are requested, but there are none
+    """
+
+    def __str__(self):
+        return "There are no tasks in the database."
+
+
+class MissingImportError(ImportError):
+    """Thrown if a library is missing.
+    """
+
+    def __init__(self, name, reason, *args, **kwargs):
+        super(MissingImportError, self).__init__(*args, **kwargs)
+        self.name = name
+        self.reason = reason
+
+    def __str__(self):
+        return "{} {}.".format(self.name, self.reason)
+
+
+class YerAnIdiotHarryError(RuntimeError):
+    """Thrown if task does not exist.
+    """
+
+    def __init__(self, name, *args, **kwargs):
+        super(YerAnIdiotHarryError, self).__init__(*args, **kwargs)
+        self.name = name
+
+    def __str__(self):
+        return "I can't believe you tried to do that, {}.".format(self.name)
+
+
+class NotSupportedError(NotImplementedError):
+    """Thrown if functionality isn't supported, and not implemented.
+    """
+
+    def __init__(self, reason, *args, **kwargs):
+        super(NotSupportedError, self).__init__(*args, **kwargs)
+        self.reason = reason
+
+    def __str__(self):
+        return self.reason
+
+
+class UnrecognizedTargetError(RuntimeError):
+    """Thrown if an unrecognized target is given for an action
+    """
+
+    def __init__(self, name, reason="", *args, **kwargs):
+        super(UnrecognizedTargetError, self).__init__(*args, **kwargs)
+        self.name = name
+        self.reason = reason
+
+    def __str__(self):
+        return "Unrecognized target {} {}.".format(self.name, self.reason)
+
+
+class UnrecognizedExecutorError(UnrecognizedTargetError):
+    """Thrown if an unrecognized executor is asked for
+    """
+
+    def __init__(self, name, reason="is not a known executor", *args, **kwargs):
+        super(UnrecognizedExecutorError, self).__init__(
+            name=name, reason=reason, *args, **kwargs
+        )
+
+
+class OutputParsingError(RuntimeError):
+    """Thrown if an output cannot be correctly parsed
+    """
+
+    def __init__(self, reason="", *args, **kwargs):
+        super(OutputParsingError, self).__init__(*args, **kwargs)
+        self.reason = reason
+
+    def __str__(self):
+        return self.reason
+
+
+class MissingDatabaseString(RuntimeError):
+    """Thrown if a database string is required and not provided
+    """
+
+    def __init__(self, reason=None, *args, **kwargs):
+        super(MissingDatabaseString, self).__init__(*args, **kwargs)
+        self.reason = reason
+
+    def __str__(self):
+        return (
+            self.reason
+            or "A database url must be defined to use a relational database. Set with qme config --database"
+        )
+
+
+class DatabaseStringFormatError(RuntimeError):
+    """Thrown if tasks are requested, but there are none
+    """
+
+    def __str__(self):
+        return (
+            "Database must start with sqlite, filesystem, mysql+pymysql, or postgres."
```

Comment:
Sqlalchemy has a defined set https://docs.sqlalchemy.org/en/13/core/engines.html#database-urls but I've only tested (and thus would want to support) a subset of that - so this is the small set that qme supports. Would it really add value to have this list be a variable somewhere else? I only really use it here - all other checks are for just single entities (e.g., if filesystem, then do this, if postgres or sqlite, do this other thing.

Reply:
If it could be anything what it supports, say it, may be with this set in e.g

Reply:
By say it, I meant say so ;-)

Reply:
But then why just not to let sqlalchemy exception to bubble up?

Reply:
Sorry I don't understand, so how would you like to see this changed?

Reply:
> But then why just not to let sqlalchemy exception to bubble up?

Because if they put in a prefix supported by sqlalechemy (but not my library) it wouldn't.

Reply:
Note, docstring seems to describe something else,... will check how you use it when get back to the laptop - tricky on the phone

Reply:
That's a bug! Fixed.